devito/operator.py

```diff
@@ -124,30 +123,29 @@ def __init__(self, expressions, **kwargs):
         # Finish instantiation
         super(Operator, self).__init__(self.name, nodes, 'int', parameters, ())
 
-    @cached_property
-    def _argument_defaults(self):
-        """
-        Derive all default values from parameters and ensure uniqueness.
-        """
-        default_args = ArgumentMap()
-        for p in self.input:
-            default_args.update(p.argument_defaults())
-        for p in self.dimensions:
-            if p.is_Sub:
-                default_args.update(p.argument_defaults(default_args))
-        return {k: default_args.reduce(k) for k in default_args}
-
     def arguments(self, **kwargs):
         """
         Process runtime arguments passed to ``.apply()` and derive
         default values for any remaining arguments.
         """
-        # First, derive all default values from parameters
-        arguments = self._argument_defaults.copy()
-
-        # Next, we insert user-provided overrides
+        arguments = {}
+        # First, we insert user-provided override
         for p in self.input + self.dimensions:
             arguments.update(p.argument_values(**kwargs))
+        # Second, derive all remaining default values from parameters
+        default_args = ArgumentMap()
+        for p in self.input:
+            if p.name not in arguments:
+                default_args.update(p.argument_defaults())
+            else:
+                default_args.update(p.argument_defaults(data=False))
+
```

Comment:
nitpicking: don't leave a blank row here because the following block of code still "belongs" to the comment "Second, derive all remaining .."